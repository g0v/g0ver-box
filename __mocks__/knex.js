import knex from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default knex;

process.setMaxListeners(0);
