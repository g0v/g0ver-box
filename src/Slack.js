import _ from 'lodash';
import slack from 'slack';

export const bot = slack.rtm.client();

const TOKEN = process.env.SLACK_TOKEN;

if (TOKEN) {
  bot.listen({ token: TOKEN });
}

bot.hello((message) => {
  console.log(`Got a message: ${message.type}`);
});

const call = (name, req = {}) => (
  new Promise((resolve, reject) => {
    const fn = _.get(slack, name);
    fn(
      { ...req, token: TOKEN },
      (err, res) => (err ? reject(err) : resolve(res))
    );
  })
);

export default {
  userInfo: async req => await call('users.info', req),
  postMessage: async req => await call('chat.postMessage', { ...req, as_user: true }),
  postMultiMessage: async (users, message) => _.map(users, user => bot.ws.send(JSON.stringify({
    id: `${user}::${Date.now()}`, type: 'message', channel: user, text: message,
  }))),
  channelInfo: async req => await call('channels.info', req),
  channelJoin: async req => await call('channels.join', req),
};
