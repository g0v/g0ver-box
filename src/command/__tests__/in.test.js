import { client } from 'knex';
import Slack from '../../Slack';
import index from '../';

describe('in command', () => {
  const data = { channel: 'D100', user: 'U03B2AB13', name: 'yutin', text: 'in' };

  describe('in [note] [hours]', () => {
    it('when no found projects', async () => {
      client.mockReturnValueOnce([]);
      client.mockReturnValueOnce([]);
      client.mockReturnValueOnce([]);

      await index(data);
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100', text: '是否為此次目標做個日誌？\n`（輸入 skip 可跳過）`',
      });

      await index({ ...data, text: 'ya! a note' });
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100', text: 'Done, in ya! a note 8 hr',
      });

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(3);
    });

    it('when found projects', async () => {
      client.mockReturnValueOnce([
        { id: '3', title: 'g0ver box' },
        { id: '5', title: 'project 001' },
      ]);

      await index(data);
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100', text: expect.stringContaining('請問是否為下列專案（輸入代碼）？'),
      });

      await index({ ...data, text: '1' });
      expect(Slack.postMessage).toHaveBeenLastCalledWith({
        channel: 'D100', text: '是否為此次目標做個日誌？\n`（輸入 skip 可跳過）`',
      });

      await index({ ...data, text: 'exit' });
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
    });

    describe('input note', () => {
      beforeEach(async () => {
        client.mockReturnValueOnce([
          { id: '3', title: 'g0ver box' },
          { id: '5', title: 'project 001' },
        ]);

        await index(data);
        await index({ ...data, text: '1' });
      });

      it('when has note', async () => {
        client.mockReturnValueOnce([]);
        client.mockReturnValueOnce(['5']);
        await index({ ...data, text: 'note information' });
        expect(Slack.postMessage).toHaveBeenLastCalledWith({
          channel: 'D100', text: 'Done, in note information 8 hr',
        });

        expect(client).toMatchSnapshot();
        expect(client).toHaveBeenCalledTimes(3);
      });

      it('when hasn\'t note', async () => {
        client.mockReturnValueOnce([]);
        client.mockReturnValueOnce(['5']);
        await index({ ...data, text: 'skip' });
        expect(Slack.postMessage).toHaveBeenLastCalledWith({
          channel: 'D100', text: 'Done, in 8 hr',
        });

        expect(client).toMatchSnapshot();
        expect(client).toHaveBeenCalledTimes(3);
      });
    });
  });

  it('in <note> [hours]', async () => {
    client.mockReturnValueOnce([]);
    await index({ ...data, text: 'in fix issue' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100', text: 'Done, in fix issue 8 hr',
    });

    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(3);
  });

  it('in <note> <hours>', async () => {
    client.mockReturnValueOnce([]);
    await index({ ...data, text: 'in fix issue 4' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100', text: 'Done, in fix issue 4 hr',
    });

    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(3);
  });
});
