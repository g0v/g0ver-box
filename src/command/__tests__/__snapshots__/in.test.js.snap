// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in command in <note> <hours> 1`] = `
Array [
  "select * from project where deleted_at is null limit 1000",
  "update task set expired_at = DATE where user_id = U03B2AB13 and deleted_at is null and expired_at > DATE",
  "insert into task (created_at, expired_at, note, updated_at) values (DATE, DATE, fix issue, DATE) returning id",
]
`;

exports[`in command in <note> [hours] 1`] = `
Array [
  "select * from project where deleted_at is null limit 1000",
  "update task set expired_at = DATE where user_id = U03B2AB13 and deleted_at is null and expired_at > DATE",
  "insert into task (created_at, expired_at, note, updated_at) values (DATE, DATE, fix issue, DATE) returning id",
]
`;

exports[`in command in [note] [hours] input note when has note 1`] = `
Array [
  "select * from project where deleted_at is null limit 1000",
  "update task set expired_at = DATE where user_id = U03B2AB13 and deleted_at is null and expired_at > DATE",
  "insert into task (created_at, expired_at, note, project_id, updated_at) values (DATE, DATE, note information, 3, DATE) returning id",
]
`;

exports[`in command in [note] [hours] input note when hasn't note 1`] = `
Array [
  "select * from project where deleted_at is null limit 1000",
  "update task set expired_at = DATE where user_id = U03B2AB13 and deleted_at is null and expired_at > DATE",
  "insert into task (created_at, expired_at, note, project_id, updated_at) values (DATE, DATE, null, 3, DATE) returning id",
]
`;

exports[`in command in [note] [hours] when found projects 1`] = `
Array [
  "select * from project where deleted_at is null limit 1000",
]
`;

exports[`in command in [note] [hours] when no found projects 1`] = `
Array [
  "select * from project where deleted_at is null limit 1000",
  "update task set expired_at = DATE where user_id = U03B2AB13 and deleted_at is null and expired_at > DATE",
  "insert into task (created_at, expired_at, note, updated_at) values (DATE, DATE, ya! a note, DATE) returning id",
]
`;
